### PR TITLE
[RLlib] Implement the SlateQ algorithm

### PR DESCRIPTION
**Pull request by desktable**
## Why are these changes needed?

Implemented the SlateQ algorithm.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/agents/slateq/slateq.py [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+# yapf: disable
+# __sphinx_doc_begin__
+DEFAULT_CONFIG = with_common_config({
+    # === Model ===
+    # Dense-layer setup for each the advantage branch and the value branch
+    # in a dueling architecture.
+    "hiddens": [256, 64, 16],
+
+    # set batchmode
+    "batch_mode": "complete_episodes",
```
Comment (sven1977) [inline rllib/agents/slateq/slateq.py:58]: Does this have to be enforced? If yes, we should add a check to `validate_config`.

File: rllib/agents/slateq/slateq.py [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+    # user/doc embedding size for the recsim environment
+    "recsim_embedding_size": 20,
+})
+# __sphinx_doc_end__
+# yapf: enable
+
+
+def validate_config(config: TrainerConfigDict) -> None:
+    """Checks the config based on settings"""
+    if config["slateq_strategy"] not in ALL_SLATEQ_STRATEGIES:
```
Comment (sven1977) [inline rllib/agents/slateq/slateq.py:131]: Can we add a framework check here for `framework==torch`? Similar to DDPPO, QMIX, etc...

File: rllib/examples/slateq.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+        raise ValueError(args.agent)
+
+    env_config = {
+        "slate_size": args.env_slate_size,
+        "seed": args.env_seed,
+        "convert_to_discrete_action_space": args.agent == "DQN",
+    }
+
+    ray.init()
+    if args.use_tune:
```
Comment (sven1977) [inline rllib/examples/slateq.py:63]: Nice! We should do this for all our example script at some point.

File: rllib/agents/slateq/slateq_torch_policy.py [hunk 4]
```diff
@@ -0,0 +1,420 @@
...
+
+        state_out = []
+        return slates_selected, state_out
+
+
+def build_slateq_model_and_distribution(
+        policy: Policy, obs_space: gym.spaces.Space,
+        action_space: gym.spaces.Space,
+        config: TrainerConfigDict) -> Tuple[ModelV2, TorchDistributionWrapper]:
+    """Build q_model and target_q_model for SlateQ
```
Comment (sven1977) [inline rllib/agents/slateq/slateq_torch_policy.py:223]: We don't have a target model.

**Review by sven1977 — APPROVED** [review]
Awesome work! Thanks. Just some minor comments and fixes to be done before merging.

File: rllib/agents/slateq/slateq.py (hunk 1 above)
Comment (desktable) [reply rllib/agents/slateq/slateq.py:58]: Only SARSA training needs this. I'll add a check in validate_config.

File: rllib/agents/slateq/slateq.py (hunk 2 above)
Comment (desktable) [reply rllib/agents/slateq/slateq.py:131]: Sure!

File: rllib/agents/slateq/slateq_torch_policy.py (hunk 4 above)
Comment (desktable) [reply rllib/agents/slateq/slateq_torch_policy.py:223]: I'll update it.

**Pull request merged**